File: leetcode/hard/count-vowels-permutation.rs
// https://leetcode.com/problems/count-vowels-permutation
//
// Recursive standard DP + Memoization approach.
// Complexity: runtime O(n), space O(n).

static VOWELS: &str = "aeiou";
const MOD: u64 = 1_000_000_007;

impl Solution {
    pub fn count_vowel_permutation(n: i32) -> i32 {
        let mut mem = vec![[0; 5]; n as _];
        (VOWELS
            .chars()
            .map(|c| Self::solve(&mut mem, c, (n - 1) as _))
            .sum::<u64>()
            % MOD) as _
    }

    fn solve(mem: &mut [[u64; 5]], c: char, n: usize) -> u64 {
        let i = VOWELS.chars().position(|x| c == x).unwrap();
        if mem[n][i] == 0 {
            mem[n][i] = if n == 0 {
                1
            } else {
                match c {
                    'a' => "e",
                    'e' => "ai",
                    'i' => "aeou",
                    'o' => "iu",
                    'u' => "a",
                    _ => unreachable!(),
                }
                .chars()
                .map(|x| Self::solve(mem, x, n - 1) % MOD)
                .sum()
            };
        }
        mem[n][i]
    }
}

File: leetcode/hard/count-vowels-permutation.cpp
// https://leetcode.com/problems/count-vowels-permutation
// Dynamic programming (bottom-up).
// Complexity: runtime O(n), space O(n).

class Solution {
  static constexpr int MOD = 1'000'000'007;

public:
  int countVowelPermutation(const int n) const {
    array<vector<int>, 5> map = {vector{1}, vector{0, 2}, vector{0, 1, 3, 4},
                                 vector{2, 4}, vector{0}};
    vector<array<int, 5>> mem(n);
    mem.front().fill(1);
    for (int i = 0; i < size(mem) - 1; ++i) {
      for (int j = 0; j < size(map); ++j) {
        for (const int k : map[j])
          mem[i + 1][k] = (mem[i + 1][k] + mem[i][j]) % MOD;
      }
    }
    int ret = 0;
    for (const int x : mem.back())
      ret = (ret + x) % MOD;
    return ret;
  }
};
